src/handlers/records-read.ts

```diff
@@ -37,15 +37,23 @@ export class RecordsReadHandler implements MethodHandler {
       return messageReplyFromError(e, 401);
     }
 
-    // get existing messages matching `recordId` so we can perform authorization
-    const query = {
-      interface : DwnInterfaceName.Records,
-      recordId  : message.descriptor.recordId
+    // get the latest active records matching the supplied filter
+    // only RecordsWrite messages will be returned due to 'isLatestBaseState' being set to true.
+    const query: Filter = {
+      interface         : DwnInterfaceName.Records,
+      isLatestBaseState : true,
+      ...Records.convertFilter(message.descriptor.filter)
     };
-    const existingMessages = await this.messageStore.query(tenant, query) as TimestampedMessage[];
+    const existingMessages = await this.messageStore.query(tenant, query) as RecordsWriteMessage[];
 
-    const newestExistingMessage = await Message.getNewestMessage(existingMessages);
+    // ensure that the returned query only contains a unique record
+    try {
+      RecordsReadHandler.enforceSingleRecordRule(existingMessages);
+    } catch (e) {
+      return messageReplyFromError(e, 400);
+    }
 
+    const newestExistingMessage = await Message.getNewestMessage(existingMessages);
     // if no record found or it has been deleted
     if (newestExistingMessage === undefined || newestExistingMessage.descriptor.method === DwnMethodName.Delete) {
```

Comment:
Another optimization as a result of filtering by `isLatestBaseState: true`
```suggestion
    if (newestExistingMessage === undefined) {
```